src/lib/Imp.hs

```diff
@@ -27,76 +26,69 @@ import Env
 import Type
 import PPrint
 import Cat
-import Subst
-import Record
 import Util (bindM2)
 
 type EmbedEnv = ([IVar], (Scope, ImpProg))
 type ImpM = Cat EmbedEnv
 
-toImpFunction :: ([ScalarTableVar], Expr) -> ImpFunction
-toImpFunction (vsIn, expr) = runImpM vsIn $ do
+toImpFunction :: ([ScalarTableVar], Block) -> ImpFunction
+toImpFunction (vsIn, block) = runImpM vsIn $ do
   (vsOut, prog) <- scopedBlock $ materializeResult
   let vsOut' = flip fmap vsOut (\(v :> IRefType t) -> v :> t)
   return $ ImpFunction vsOut' vsIn prog
   where
     materializeResult = do
-      ans <- toImpExpr mempty expr
-      (outDest, vsOut) <- allocDest Unmanaged $ getType expr
+      ans <- toImpBlock mempty block
+      (outDest, vsOut) <- allocDest Unmanaged $ getType block
       copyAtom outDest ans
       return vsOut
 
 runImpM :: [ScalarTableVar] -> ImpM a -> a
 runImpM inVars m = fst $ runCat m (mempty, (inVarScope, mempty))
   where
     inVarScope :: Scope
-    inVarScope = foldMap varAsEnv $ fmap (fmap $ const ()) inVars
+    inVarScope = fmap (const Nothing) $
+      foldMap varAsEnv $ fmap (fmap $ const ()) inVars
```

Comment:
Can't you replace `const ()` with `const Nothing` and skip the outer `fmap` here?

Reply:
Yep, good one